grblas/scalar.py

```diff
@@ -17,18 +18,68 @@ def __repr__(self):
-            return self.value == other.value
-        else:
-            return self.value == other
+        return self.isequal(other)
 
     def __bool__(self):
         if self.is_empty:
             return False
         return bool(self.value)
 
+    def isequal(self, other, *, check_dtype=False):
+        """
+        Check for exact equality
+
+        If `check_dtype` is True, also checks that dtypes match
+        For equality of floating point Vectors, consider using `isclose`
+        """
+        if not isinstance(other, Scalar):
+            if other is None:
+                return self.is_empty
+            try:
+                # Check if other is a literal scalar, which we should handle
+                dtype = dtypes.lookup(type(other))
+            except KeyError:
+                raise TypeError(f'Argument of isequal must be a known scalar type, not {type(other)}')
+            other = Scalar.from_value(other, dtype=dtype)
+            # Don't check dtype if we had to infer dtype of `other`
+            check_dtype = False
+        if check_dtype and self.dtype != other.dtype:
+            return False
+        if self.is_empty or other.is_empty:
+            return self.is_empty is other.is_empty
+        # For now, compare values in Python.  We can get more sophisticated
+        # if there is a need by converting both scalars to 1-d Vectors.
+        # Hopefully scalar types will be added to the GraphBLAS spec.
+        return self.value == other.value
+
+    def isclose(self, other, *, rel_tol=1e-7, abs_tol=0.0, check_dtype=False):
+        """
+        Check for approximate equality
+
+        If `check_dtype` is True, also checks that dtypes match
+        Closeness check is equivalent to `abs(a-b) <= max(rel_tol * max(abs(a), abs(b)), abs_tol)`
+        """
+        if not isinstance(other, Scalar):
+            if other is None:
+                return self.is_empty
+            try:
+                # Check if other is a literal scalar, which we should handle
+                dtype = dtypes.lookup(type(other))
+            except KeyError:
+                raise TypeError(f'Argument of isclose must be a known scalar type, not {type(other)}')
+            other = Scalar.from_value(other, dtype=dtype)
+            # Don't check dtype if we had to infer dtype of `other`
+            check_dtype = False
+        if check_dtype and self.dtype != other.dtype:
+            return False
+        if self.is_empty or other.is_empty:
+            return self.is_empty is other.is_empty
+        # We can't yet call a UDF on a scalar, so lets convert to 1-d vector
+        left = grblas.vector.Vector.from_values([0], [self.value], size=1, dtype=self.dtype)
+        right = grblas.vector.Vector.from_values([0], [other.value], size=1, dtype=other.dtype)
+        matches = left.ewise_mult(right, grblas.binary.isclose(rel_tol, abs_tol)).new(dtype=bool)
+        return matches.reduce(grblas.monoid.land).value
```

Comment:
Is this actually faster than just calling `math.isclose(self.value, other.value, rel_tol=rel_tol, abs_tol=abs_tol)`?
I agree this is a good approach once GrB_Scalar exists, but for now, this seems overly complicated.

Reply:
No idea about performance.  Yes, it is complicated, but I trust it to be consistent.  I don't know if `math.isclose` will behave the same or not.

Reply:
Fair enough

Reply:
Oh yeah, I also want to expose the compiled numba function from UDFs.  Once we have this, I want to simplify this to something like
```python
grblas.binary.isclose.numba_func(self.value, other.value)
```